File: ZHDragView/zhdraglib/src/main/java/com/zihuan/recyclerview/drag/DragItemTouchHelper.java
package com.zihuan.recyclerview.drag;

import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import java.util.Collections;
import java.util.List;

public class DragItemTouchHelper extends ItemTouchHelper.Callback {

    RecyclerView.Adapter mAdapter;
    List mList;
    DragItemTouchInterface mTouchInterface;

    public DragItemTouchHelper(RecyclerView.Adapter adapter, List list) {
        mAdapter = adapter;
        mList = list;
    }

    /**
     * @param adapter
     * @param list
     * @param touchInterface 回调接口
     */
    public DragItemTouchHelper(RecyclerView.Adapter adapter, List list, DragItemTouchInterface touchInterface) {
        mAdapter = adapter;
        mList = list;
        mTouchInterface = touchInterface;
    }

    //  设置是否滑动时间，以及拖拽的方向
    //  如果是列表布局的话则拖拽方向有DOWN和UP，
    //  如果是网格布局的话有DOWN和UP和LEFT和RIGHT4个方向
    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        if (recyclerView.getLayoutManager() instanceof GridLayoutManager) {
            final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
            final int swipeFlags = 0;
            return makeMovementFlags(dragFlags, swipeFlags);
        } else {
            final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            final int swipeFlags = 0;
            return makeMovementFlags(dragFlags, swipeFlags);
        }
    }

    //    正在拖动中
    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        //得到当拖拽的viewHolder的Position
        int fromPosition = viewHolder.getAdapterPosition();
        //拿到当前拖拽到的item的viewHolder
        int toPosition = target.getAdapterPosition();
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(mList, i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(mList, i, i - 1);
            }
        }
        mAdapter.notifyItemMoved(fromPosition, toPosition);
        return true;
    }

    //拖拽完成
    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
    }

    //    选中view
    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        super.onSelectedChanged(viewHolder, actionState);
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE && mTouchInterface != null)
            mTouchInterface.onSelectedChanged(viewHolder, actionState);
    }

    //    放开view
    @Override
    public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);
        if (mTouchInterface != null)
            mTouchInterface.clearView(recyclerView, viewHolder);
    }

    //重写拖拽不可用
    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

}

File: ZHDragView/library/src/main/java/com/zihuan/recyclerview/drag/DragItemTouchHelper.kt
package com.zihuan.recyclerview.drag

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import java.util.*

/**
 * 在单个RecyclerView中拖拽Item
 */
class DragItemTouchHelper : ItemTouchHelper.Callback {
    var mAdapter: RecyclerView.Adapter<*>
    var mList: List<*>
    var mTouchInterface: DragItemTouchInterface? = null

    constructor(adapter: RecyclerView.Adapter<*>, list: List<*>) {
        mAdapter = adapter
        mList = list
    }

    /**
     * @param adapter
     * @param list
     * @param touchInterface 回调接口
     */
    constructor(adapter: RecyclerView.Adapter<*>, list: List<*>, touchInterface: DragItemTouchInterface?) {
        mAdapter = adapter
        mList = list
        mTouchInterface = touchInterface
    }

    //  设置是否滑动时间，以及拖拽的方向
//  如果是列表布局的话则拖拽方向有DOWN和UP，
//  如果是网格布局的话有DOWN和UP和LEFT和RIGHT4个方向
    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: ViewHolder): Int {
        return if (recyclerView.layoutManager is GridLayoutManager) {
            val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
            val swipeFlags = 0
            makeMovementFlags(dragFlags, swipeFlags)
        } else {
            val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
            val swipeFlags = 0
            makeMovementFlags(dragFlags, swipeFlags)
        }
    }

    //    正在拖动中
    override fun onMove(recyclerView: RecyclerView, viewHolder: ViewHolder, target: ViewHolder): Boolean { //得到当拖拽的viewHolder的Position
        val fromPosition = viewHolder.adapterPosition
        //拿到当前拖拽到的item的viewHolder
        val toPosition = target.adapterPosition
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                Collections.swap(mList, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                Collections.swap(mList, i, i - 1)
            }
        }
        mAdapter.notifyItemMoved(fromPosition, toPosition)
        return true
    }

    //拖拽完成
    override fun onSwiped(viewHolder: ViewHolder, direction: Int) {}

    //    选中view
    override fun onSelectedChanged(viewHolder: ViewHolder?, actionState: Int) {
        super.onSelectedChanged(viewHolder, actionState)
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE && mTouchInterface != null) mTouchInterface!!.onSelectedChanged(viewHolder, actionState)
    }

    //    放开view
    override fun clearView(recyclerView: RecyclerView, viewHolder: ViewHolder) {
        super.clearView(recyclerView, viewHolder)
        if (mTouchInterface != null) mTouchInterface!!.clearView(recyclerView, viewHolder)
    }

    //重写拖拽不可用
    override fun isLongPressDragEnabled(): Boolean {
        return false
    }
}